modules/mbedtls_curl/http_client_curl.cpp

```diff
@@ -0,0 +1,440 @@
+	// that can be associated with request handlers. It's used here to
+	// keep track of certain data that needs to be manipulated throughout
+	// the pipeline.
+	// @see https://curl.se/libcurl/c/CURLOPT_PRIVATE.html
+	curl_easy_setopt(eh, CURLOPT_PRIVATE, ctx);
+
+	curl_multi_add_handle(curl, eh);
+
+	in_flight = true;
+	status = STATUS_REQUESTING;
+	still_running = 0;
+	return OK;
+}
+
+Error HTTPClientCurl::get_response_headers(List<String> *r_response) {
+	*r_response = response_headers;
+	return OK;
+}
+
+Error HTTPClientCurl::connect_to_host(const String &p_host, int p_port, bool p_ssl, bool p_verify_host) {
+	if (curl == nullptr) {
+		curl = curl_multi_init();
+	}
+
+	response_code = 0;
+	body_size = -1;
+	response_headers.clear();
+	response_available = false;
+	response_code = 0;
+	response_chunks.clear();
+	keep_alive = false;
+	status = STATUS_CONNECTED;
+	scheme = p_host.begins_with("https://") ? "https://" : "http://";
+	host = p_host.trim_prefix("http://").trim_prefix("https://");
+	port = p_port;
+	ssl = p_ssl;
+	verify_host = p_verify_host;
+
+	return OK;
+}
+
+HTTPClientCurl::~HTTPClientCurl() {
+	close();
+}
+
+void HTTPClientCurl::close() {
+	if (curl) {
+		curl_multi_cleanup(curl);
+		curl = nullptr;
+	}
+}
+
+Error HTTPClientCurl::request(Method p_method, const String &p_url, const Vector<String> &p_headers, const uint8_t *p_body, int p_body_size) {
+	// Only one request can be in flight at a time.
+	if (in_flight) {
+		return ERR_ALREADY_IN_USE;
+	}
+
+	method = p_method;
+	url = p_url;
+	request_headers = p_headers;
+	request_body = p_body;
+	request_body_size = p_body_size;
+
+	if (host.is_valid_ip_address()) {
+		_request(host, false);
+	} else {
+		resolver_id = IP::get_singleton()->resolve_hostname_queue_item(host, IP::Type::TYPE_ANY);
+		status = STATUS_RESOLVING;
+	}
+
+	return OK;
+}
+
+Error HTTPClientCurl::poll() {
+	if (status == STATUS_RESOLVING) {
+		ERR_FAIL_COND_V(resolver_id == IP::RESOLVER_INVALID_ID, ERR_BUG);
+		return _resolve_dns();
+	}
+
+	// Important! Since polling libcurl will greedily read response data from the
+	// network we don't want to poll when we are in STATUS_BODY state. The reason
+	// for this is that the HTTPClient API is expected to only read from the network
+	// when read_response_body_chunk is called. This means that here, in poll, we only
+	// poll libcurl when we are not in the STATUS_BODY state and we poll libcurl in
+	// read_response_body_chunk instead, when we are in STATUS_BODY state.
+	if (status != STATUS_BODY) {
+		return _poll_curl();
+	}
+	return OK;
+}
+
+PackedByteArray HTTPClientCurl::read_response_body_chunk() {
+	if (status == STATUS_BODY) {
+		Error err = _poll_curl();
+		if (err != OK) {
+			ERR_PRINT_ONCE("Failed when polling curl in STATUS_BODY. RC: " + String::num_int64(err));
+			return PackedByteArray();
+		}
+	}
+	if (response_chunks.is_empty()) {
+		status = keep_alive ? STATUS_CONNECTED : STATUS_DISCONNECTED;
```

Comment:
So the big problem with this is that between one `_poll_curl()` and the other there might not have been any new data received, which will be interpreted as the request being over.

Reply:
Alright @Faless I'm not terribly happy with my solution to this, but it's the best I could think of after several hours of racking my brain. I left a comment in the code describing the problem and the up- and downsides to my solution.